### PR TITLE
Restructuring middleware docs pages

### LINKED ISSUE #840: Add HowTos for Middleware HTTP components
Only two of the middleware components have documentation currently. These are the components
https://github.com/dapr/components-contrib/tree/master/middleware/http

Only OPA and OAUTH have docs here https://github.com/dapr/docs/tree/master/howto#middleware
Need to add "HowTo" docs for the other four to explain their usage.

- bearer
- nethttpadaptor
**- oauth2**
- oauth2clientcredentials
**- opa**
- ratelimit

Note; Need to do ensure that the middleware docs reference these which are confusing to find currently.
https://v1-rc2.docs.dapr.io/operations/security/oauth/
https://v1-rc2.docs.dapr.io/developing-applications/middleware/middleware-opa-policies/

### PR DESCRIPTION
**Pull request by pkedy**
Restructuring the middleware docs pages to be consistent with the other sections of the docs (e.g. Supported secret stores).

Resolves #840, #1076, and #836

The middleware overview page still needs content. Much of that feels like it could already be on the middleware-concept page.

**Checklist:**

- [x] [Read the contribution guide](https://docs.dapr.io/contributing/contributing-docs/)
- [x] Commands include options for Linux, MacOS, and Windows within codetabs
- [x] New file and folder names are globally unique
- [x] Page references use shortcodes instead of markdown or URL links
- [x] Images use HTML style and have alternative text
- [x] Places where multiple code/command options are given have codetabs

### CODE REVIEW

**Review by msfussell — APPROVED** [review]
Look GREAT! Thanks for doing this.

File: daprdocs/content/en/developing-applications/middleware/middleware-overview.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    }
+  }
+}
+```
+
+## Adding new middleware components
+
+Your middleware component can be contributed to the [components-contrib repository](https://github.com/dapr/components-contrib/tree/master/middleware). 
+
+After the components-contrib has been accepted, submit another pull request against the [Dapr runtime repository](https://github.com/dapr/dapr) to register the new middleware type. You'll need to modify **[runtime.WithHTTPMiddleware](https://github.com/dapr/dapr/blob/f4d50b1369e416a8f7b93e3e226c4360307d1313/cmd/daprd/main.go#L394-L424)** method in [cmd/daprd/main.go](https://github.com/dapr/dapr/blob/master/cmd/daprd/main.go) to register your middleware with Dapr's runtime.
```
Comment (mukundansundar) [inline daprdocs/content/en/developing-applications/middleware/middleware-overview.md]:
```suggestion
After the components-contrib change has been accepted, submit another pull request against the [Dapr runtime repository](https://github.com/dapr/dapr) to register the new middleware type. You'll need to modify **[runtime.WithHTTPMiddleware](https://github.com/dapr/dapr/blob/f4d50b1369e416a8f7b93e3e226c4360307d1313/cmd/daprd/main.go#L394-L424)** method in [cmd/daprd/main.go](https://github.com/dapr/dapr/blob/master/cmd/daprd/main.go) to register your middleware with Dapr's runtime.
```

**Review by mukundansundar — COMMENTED** [review]
@msfussell @pkedy Should these files be moved under `operations/components/middleware`? instead of `developing-application/middleware`?

File: daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-oauth2.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+| tokenURL       | The endpoint is used by the client to obtain an access token by presenting its authorization grant or refresh token                                                          | `"https://accounts.google.com/o/oauth2/token"`     |
+| redirectURL    | The URL of your web application that the authorization server should redirect to once the user has authenticated                                                             | `"https://myapp.com"`                              |
+| authHeaderName | The authorization header name to forward to your application                                                                                                                 | `"authorization"`                                |
+| forceHTTPS     | If true, enforces the use of TLS/SSL                                                                                                                                         | `"true"`,`"false"`                                           |
+
+## Dapr configuration
```
Comment (mukundansundar) [inline daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-oauth2.md:51]: this is a repetition that's there in every middleware component. Can this be moved to middleware-overview with a link from this page referencing that as the way to apply middleware configuration? Similar to how it is done for all the other different components.

File: daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-oauth2clientcredentials.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+| endpointParamsQuery | Specifies additional parameters for requests to the token endpoint                                                                                                                                                                                                                                                                                                                                                                                                                                  | `true`                                             |
+| authStyle           | Optionally specifies how the endpoint wants the client ID & client secret sent. See the table of possible values below. | `0`                                                |
```
Comment (mukundansundar) [inline daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-oauth2clientcredentials.md]: nitpick 
```suggestion
| authStyle           | Optionally specifies how the endpoint wants the client ID & client secret sent. See the table of possible values below | `0`                                                |
```

File: daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-oauth2clientcredentials.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+**Possible values for `authStyle`**
```
Comment (mukundansundar) [inline daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-oauth2clientcredentials.md]:
```suggestion
### AuthStyle
```
Add as a heading and add a link in the description also?

File: daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-opa.md [hunk 5]
```diff
@@ -69,6 +68,30 @@ spec:
...
+|-----------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------|
+| rego            | The Rego policy language                                                                                                                                                                              | See above                                                         |
+| defaultStatus   | The status code to return for denied responses                                                                                                                                                        | `"https://accounts.google.com"`, `"https://login.salesforce.com"` |
+| includedHeaders | A comma-separated set of case-insensitive headers to include in the request input. Request headers are not passed to the policy by default. Include to receive incoming request headers in the input. | `"x-my-custom-header, x-jwt-header"`                              |
```
Comment (mukundansundar) [inline daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-opa.md]: nitpick
```suggestion
| includedHeaders | A comma-separated set of case-insensitive headers to include in the request input. Request headers are not passed to the policy by default. Include to receive incoming request headers in the input | `"x-my-custom-header, x-jwt-header"`                              |
```

File: daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-rate-limit.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  metadata:
+  - name: maxRequestsPerSecond
+    value: 10
+```
+
+## Spec metadata fields
+
+| Field       | Details                                                                                                                                                                              | Example |
+|----------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------|
+| maxRequestsPerSecond | The maximum requests per second by remote IP and path. Something to consider is that **the limit is enforced independently in each Dapr sidecar and not cluster wide.** | `10`    |
```
Comment (mukundansundar) [inline daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-rate-limit.md]: nipick
```suggestion
| maxRequestsPerSecond | The maximum requests per second by remote IP and path. Something to consider is that **the limit is enforced independently in each Dapr sidecar and not cluster wide** | `10`    |
```

File: daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-rate-limit.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+To be applied, the middleware must be referenced in [configuration]({{< ref configuration-concept.md >}}). See [middleware pipelines]({{< ref "middleware-concept.md#customize-processing-pipeline">}}).
+
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: Configuration
+metadata:
+  name: daprConfig
+spec:
+  tracing:
```
Comment (mukundansundar) [inline daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-rate-limit.md]: why is tracing here in this config, while not there in other configs? It is not needed for rate limit so why is it shown?

**Review by mukundansundar — COMMENTED** [review]
@pkedy Looks awesome. Just left some comments and suggestions.

File: daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-oauth2clientcredentials.md (hunk 4 above)
Comment (pkedy) [reply daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-oauth2clientcredentials.md]: I made it a `###` header but kept the text to preserve the capitalization of the field name.

File: daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-oauth2.md (hunk 2 above)
Comment (pkedy) [reply daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-oauth2.md:51]: If you mean the `Dapr configuration` section, there is already content in the overview page for that. I think having it retrofitted in each middleware page is just a convenience thing. I'm leaving this for @msfussell to decide. :)

File: daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-rate-limit.md (hunk 7 above)
Comment (pkedy) [reply daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-rate-limit.md]: I removed this

**Comment by AaronCrawfis** [comment]
> @msfussell @pkedy Should these files be moved under `operations/components/middleware`? instead of `developing-application/middleware`?

My vote is developing-applications. It's too closely tied with the functionality and design of the app to be placed under operations. We should leave operations for deployment/infrastructure topics

File: daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-oauth2.md (hunk 2 above)
Comment (msfussell) [reply daprdocs/content/en/developing-applications/middleware/supported-middleware/middleware-oauth2.md:51]: Yes, this is just convenience thing. Middleware is slightly different in that you need to reference this in config and this is a reminder. Let's keep this for now and we can PR this out later if needed.

**Comment by msfussell** [comment]
> > @msfussell @pkedy Should these files be moved under `operations/components/middleware`? instead of `developing-application/middleware`?
> 
> My vote is developing-applications. It's too closely tied with the functionality and design of the app to be placed under operations. We should leave operations for deployment/infrastructure topics

Let's PR this in under developing-apps for now and then see the structure. Maybe we can do to this what we intend to do with Components which is move the actual components definitions under references with another PR and link them.

**Comment by msfussell** [comment]
Unless there are other comments I am going to merge this PR?

**Pull request merged**